network/gossip/libp2p/peerManager_test.go

```diff
@@ -114,108 +115,128 @@ func (ts *PeerManagerTestSuite) TestUpdatePeers() {
 
 		pm.updatePeers()
 
-		connector.AssertNumberOfCalls(ts.T(), "ConnectPeers", 4)
-		connector.AssertNumberOfCalls(ts.T(), "DisconnectPeers", 2)
+		connector.AssertNumberOfCalls(suite.T(), "ConnectPeers", 4)
+		connector.AssertNumberOfCalls(suite.T(), "DisconnectPeers", 2)
 	})
 }
 
-// TestPeriodicPeerUpdate tests that the peermanager runs periodically
-func (ts *PeerManagerTestSuite) TestPeriodicPeerUpdate() {
+// TestPeriodicPeerUpdate tests that the peer manager runs periodically
+func (suite *PeerManagerTestSuite) TestPeriodicPeerUpdate() {
 	currentIDs := unittest.IdentityListFixture(10)
 	idProvider := func() (flow.IdentityList, error) {
 		return currentIDs, nil
 	}
 
 	connector := new(mock.Connector)
-	connector.On("ConnectPeers", ts.ctx, testifymock.Anything).Return(nil)
-	connector.On("DisconnectPeers", ts.ctx, testifymock.Anything).Return(nil)
-	pm := NewPeerManager(ts.ctx, ts.log, idProvider, connector)
+	wg := &sync.WaitGroup{} // keeps track of number of calls on `ConnectPeers`
+	count, times := 0, 2    // we expect it to be called twice at least
+	wg.Add(times)
+	connector.On("ConnectPeers", suite.ctx, testifymock.Anything).Run(func(args testifymock.Arguments) {
+		if count < times {
+			count++
+			wg.Done()
+		}
+	}).Return(nil)
+	connector.On("DisconnectPeers", suite.ctx, testifymock.Anything).Return(nil)
+	pm := NewPeerManager(suite.ctx, suite.log, idProvider, connector)
 
 	PeerUpdateInterval = 5 * time.Millisecond
 	err := pm.Start()
-	assert.NoError(ts.T(), err)
-	assert.Eventually(ts.T(), func() bool {
-		return connector.AssertNumberOfCalls(ts.T(), "ConnectPeers", 2)
-	}, 2*PeerUpdateInterval+4*time.Millisecond, 2*PeerUpdateInterval)
```

Comment:
**A root cause of flakiness**: in this test we are checking whether we receive _at least two_ calls on `ConnectPeers` or not? However, the syntax checks for the _exact two_ calls on it. Due to the asynchrony among `assert.Eventually` and peer update goroutine, it was sometimes the case that the method was getting called 3 times, and hence would fail the test.